File: src/agents-python/services/financial_service.py
"""
Financial analysis service - Core business logic for financial operations, reporting, and analytics.
"""
import asyncio
from typing import List, Dict, Any, Optional, Tuple
from datetime import datetime, timedelta
from decimal import Decimal
import random

from agents_python.financial_models import (
    SalesReport, RevenueData, ExpenseRecord, EmployeePerformance, 
    InventoryData, CustomerData, FinancialMetrics, TrendDirection,
    ReportType, DataSourceType
)


class FinancialService:
    """Core business logic for financial analysis, reporting, and business metrics."""
    
    def __init__(self):
        """Initialize the financial service with mocked data."""
        self._initialize_mock_data()
    
    def _initialize_mock_data(self):
        """Initialize mock data for demonstration purposes."""
        # This would typically connect to databases/APIs in a real implementation
        pass
    
    async def get_sales_reports(self, date_range: Optional[str] = None, category: Optional[str] = None) -> List[SalesReport]:
        """Get sales reports with optional filtering by date range and category."""
        await asyncio.sleep(0.1)  # Simulate async operation
        
        sales_data = [
            SalesReport(
                product_id="PROD001",
                product_name="Enterprise Software License",
                category="Software",
                sales_amount=Decimal("125000.00"),
                units_sold=50,
                revenue=Decimal("125000.00"),
                profit_margin=0.75,
                sales_date=datetime.now() - timedelta(days=15),
                region="North America",
                sales_rep="Alice Johnson"
            ),
            SalesReport(
                product_id="PROD002",
                product_name="Consulting Services",
                category="Services",
                sales_amount=Decimal("89500.00"),
                units_sold=179,
                revenue=Decimal("89500.00"),
                profit_margin=0.65,
                sales_date=datetime.now() - timedelta(days=8),
                region="Europe",
                sales_rep="Bob Smith"
            ),
            SalesReport(
                product_id="PROD003",
                product_name="Hardware Components",
                category="Hardware",
                sales_amount=Decimal("67800.00"),
                units_sold=226,
                revenue=Decimal("67800.00"),
                profit_margin=0.35,
                sales_date=datetime.now() - timedelta(days=22),
                region="Asia Pacific",
                sales_rep="Carol Chen"
            ),
            SalesReport(
                product_id="PROD004",
                product_name="Training Package",
                category="Services",
                sales_amount=Decimal("34500.00"),
                units_sold=115,
                revenue=Decimal("34500.00"),
                profit_margin=0.80,
                sales_date=datetime.now() - timedelta(days=5),
                region="North America",
                sales_rep="David Wilson"
            ),
            SalesReport(
                product_id="PROD005",
                product_name="Mobile App License",
                category="Software",
                sales_amount=Decimal("45200.00"),
                units_sold=904,
                revenue=Decimal("45200.00"),
                profit_margin=0.85,
                sales_date=datetime.now() - timedelta(days=12),
                region="Europe",
                sales_rep="Eva Rodriguez"
            )
        ]
        
        # Apply filters
        filtered_data = sales_data
        if category:
            filtered_data = [s for s in filtered_data if s.category.lower() == category.lower()]
        
        return filtered_data
    
    async def get_revenue_data(self, period: Optional[str] = None) -> List[RevenueData]:
        """Get revenue data by period."""
        await asyncio.sleep(0.1)
        
        revenue_data = [
            RevenueData(
                period="Q4 2024",
                revenue=Decimal("2850000.00"),
                recurring_revenue=Decimal("2100000.00"),
                one_time_revenue=Decimal("750000.00"),
                growth_rate=12.5,
                revenue_source="Software Licenses",
                department="Sales"
            ),
            RevenueData(
                period="Q3 2024",
                revenue=Decimal("2540000.00"),
                recurring_revenue=Decimal("1950000.00"),
                one_time_revenue=Decimal("590000.00"),
                growth_rate=8.3,
                revenue_source="Services",
                department="Consulting"
            ),
            RevenueData(
                period="Q2 2024",
                revenue=Decimal("2345000.00"),
                recurring_revenue=Decimal("1800000.00"),
                one_time_revenue=Decimal("545000.00"),
                growth_rate=15.2,
                revenue_source="Hardware Sales",
                department="Sales"
            ),
            RevenueData(
                period="Q1 2024",
                revenue=Decimal("2038000.00"),
                recurring_revenue=Decimal("1650000.00"),
                one_time_revenue=Decimal("388000.00"),
                growth_rate=6.7,
                revenue_source="Training",
                department="Education"
            )
        ]
        
        if period:
            revenue_data = [r for r in revenue_data if period.lower() in r.period.lower()]
        
        return revenue_data
    
    async def get_employee_performance(self, department: Optional[str] = None) -> List[EmployeePerformance]:
        """Get employee performance metrics."""
        await asyncio.sleep(0.1)
        
        performance_data = [
            EmployeePerformance(
                employee_id="EMP001",
                employee_name="Alice Johnson",
                department="Sales",
                role="Senior Sales Manager",
                performance_score=4.7,
                sales_target=Decimal("500000.00"),
                sales_achieved=Decimal("625000.00"),
                commission=Decimal("31250.00"),
                review_period="Q4 2024"
            ),
            EmployeePerformance(
                employee_id="EMP002",
                employee_name="Bob Smith",
                department="Consulting",
                role="Principal Consultant",
                performance_score=4.5,
                sales_target=Decimal("300000.00"),
                sales_achieved=Decimal("320000.00"),
                commission=Decimal("16000.00"),
                review_period="Q4 2024"
            ),
            EmployeePerformance(
                employee_id="EMP003",
                employee_name="Carol Chen",
                department="Engineering",
                role="Software Engineer",
                performance_score=4.8,
                review_period="Q4 2024"
            )
        ]
        
        if department:
            performance_data = [p for p in performance_data if p.department.lower() == department.lower()]
        
        return performance_data
    
    async def get_inventory_data(self, category: Optional[str] = None) -> List[InventoryData]:
        """Get inventory levels and data."""
        await asyncio.sleep(0.1)
        
        inventory_data = [
            InventoryData(
                item_id="INV001",
                item_name="Server Unit Model X",
                category="Hardware",
                current_stock=45,
                reorder_level=20,
                unit_cost=Decimal("2500.00"),
                retail_price=Decimal("3750.00"),
                supplier="TechCorp Inc",
                last_updated=datetime.now() - timedelta(days=2),
                location="Warehouse A"
            ),
            InventoryData(
                item_id="INV002",
                item_name="Software License Keys",
                category="Software",
                current_stock=500,
                reorder_level=100,
                unit_cost=Decimal("150.00"),
                retail_price=Decimal("299.00"),
                supplier="SoftDev Solutions",
                last_updated=datetime.now() - timedelta(hours=6),
                location="Digital Inventory"
            ),
            InventoryData(
                item_id="INV003",
                item_name="Network Equipment",
                category="Hardware",
                current_stock=12,
                reorder_level=15,
                unit_cost=Decimal("1200.00"),
                retail_price=Decimal("1899.00"),
                supplier="NetWork Pro",
                last_updated=datetime.now() - timedelta(days=1),
                location="Warehouse B"
            )
        ]
        
        if category:
            inventory_data = [i for i in inventory_data if i.category.lower() == category.lower()]
        
        return inventory_data
    
    async def get_customer_data(self, segment: Optional[str] = None) -> List[CustomerData]:
        """Get customer demographics and behavior analytics."""
        await asyncio.sleep(0.1)
        
        customer_data = [
            CustomerData(
                customer_id="CUST001",
                acquisition_date=datetime.now() - timedelta(days=180),
                acquisition_cost=Decimal("5000.00"),
                lifetime_value=Decimal("125000.00"),
                segment="enterprise",
                region="North America",
                total_orders=15,
                avg_order_value=Decimal("8333.33"),
                last_order_date=datetime.now() - timedelta(days=10),
                churn_risk="low"
            ),
            CustomerData(
                customer_id="CUST002",
                acquisition_date=datetime.now() - timedelta(days=90),
                acquisition_cost=Decimal("1500.00"),
                lifetime_value=Decimal("45000.00"),
                segment="smb",
                region="Europe",
                total_orders=8,
                avg_order_value=Decimal("5625.00"),
                last_order_date=datetime.now() - timedelta(days=25),
                churn_risk="medium"
            ),
            CustomerData(
                customer_id="CUST003",
                acquisition_date=datetime.now() - timedelta(days=45),
                acquisition_cost=Decimal("200.00"),
                lifetime_value=Decimal("2400.00"),
                segment="individual",
                region="Asia Pacific",
                total_orders=12,
                avg_order_value=Decimal("200.00"),
                last_order_date=datetime.now() - timedelta(days=3),
                churn_risk="low"
            )
        ]
        
        if segment:
            customer_data = [c for c in customer_data if c.segment.lower() == segment.lower()]
        
        return customer_data
    
    async def calculate_financial_metrics(self, metric_names: List[str], period: Optional[str] = None) -> List[FinancialMetrics]:
        """Calculate various financial metrics."""
        await asyncio.sleep(0.2)  # Simulate computation time
        
        # Mock calculations - in real implementation, these would be complex financial calculations
        metrics = []
        
        if "customer_acquisition_cost" in metric_names:
            metrics.append(FinancialMetrics(
                metric_name="Customer Acquisition Cost",
                value=Decimal("2567.00"),
                period=period or "Q4 2024",
                calculation_method="Total Marketing Spend / New Customers Acquired",
                benchmark=Decimal("3000.00"),
                trend=TrendDirection.DOWN,  # Lower CAC is better
                metadata={
                    "marketing_spend": "385000",
                    "new_customers": 150,
                    "improvement_vs_last_quarter": "15%"
                }
            ))
        
        if "revenue_growth_rate" in metric_names:
            metrics.append(FinancialMetrics(
                metric_name="Revenue Growth Rate",
                value=Decimal("12.5"),
                period=period or "Q4 2024",
                calculation_method="((Current Period Revenue - Previous Period Revenue) / Previous Period Revenue) * 100",
                benchmark=Decimal("10.0"),
                trend=TrendDirection.UP,
                metadata={
                    "current_revenue": "2850000",
                    "previous_revenue": "2540000",
                    "target_growth": "10%"
                }
            ))
        
        if "profit_margin" in metric_names:
            metrics.append(FinancialMetrics(
                metric_name="Gross Profit Margin",
                value=Decimal("68.5"),
                period=period or "Q4 2024",
                calculation_method="((Revenue - Cost of Goods Sold) / Revenue) * 100",
                benchmark=Decimal("65.0"),
                trend=TrendDirection.UP,
                metadata={
                    "revenue": "2850000",
                    "cogs": "897750",
                    "industry_average": "62%"
                }
            ))
        
        return metrics
    
    async def analyze_trends(self, data_type: str, period: str = "6months") -> Dict[str, Any]:
        """Analyze trends in financial data."""
        await asyncio.sleep(0.3)
        
        # Mock trend analysis - in real implementation, this would use pandas/numpy for statistical analysis
        trend_data = {
            "sales": {
                "direction": "up",
                "percentage_change": 18.7,
                "key_drivers": ["New product launches", "Market expansion", "Improved sales process"],
                "seasonal_patterns": "Q4 typically shows 25% increase due to enterprise budget cycles",
                "forecast": "Continued growth expected with 15% increase in Q1 2025",
                "data_points": [
                    {"period": "2024-07", "value": 2100000},
                    {"period": "2024-08", "value": 2250000},
                    {"period": "2024-09", "value": 2380000},
                    {"period": "2024-10", "value": 2450000},
                    {"period": "2024-11", "value": 2600000},
                    {"period": "2024-12", "value": 2850000}
                ]
            }
        }
        
        return trend_data.get(data_type, {"error": f"No trend data available for {data_type}"})
    
    async def search_financial_data(self, query: str, data_sources: Optional[List[str]] = None) -> Dict[str, Any]:
        """Search across financial data sources."""
        await asyncio.sleep(0.2)
        
        # Mock search functionality
        results = {
            "query": query,
            "sources_searched": data_sources or ["sales", "revenue", "expenses", "customers"],
            "total_results": random.randint(5, 25),
            "results": []
        }
        
        # Simulate search results based on query
        if "top" in query.lower() and "product" in query.lower():
            results["results"] = [
                {
                    "source": "sales_reports",
                    "title": "Enterprise Software License - Top Performer Q4",
                    "summary": "Generated $125,000 in revenue with 75% profit margin",
                    "metric_value": 125000,
                    "relevance_score": 0.95
                },
                {
                    "source": "sales_reports", 
                    "title": "Consulting Services - Strong Growth",
                    "summary": "Achieved $89,500 in sales with consistent demand",
                    "metric_value": 89500,
                    "relevance_score": 0.88
                }
            ]
        elif "trend" in query.lower():
            results["results"] = [
                {
                    "source": "revenue_analysis",
                    "title": "6-Month Revenue Trend Analysis",
                    "summary": "18.7% growth trend with strong Q4 performance",
                    "metric_value": 18.7,
                    "relevance_score": 0.92
                }
            ]
        
        return results
    
    def generate_mock_chart_data(self, chart_type: str, data_series: List[str]) -> Dict[str, Any]:
        """Generate mock data for chart visualization."""
        # This would typically integrate with plotting libraries like matplotlib/plotly
        return {
            "chart_type": chart_type,
            "data": {
                "labels": ["Jan", "Feb", "Mar", "Apr", "May", "Jun"],
                "datasets": [
                    {
                        "label": series,
                        "data": [random.randint(1000, 5000) for _ in range(6)]
                    } for series in data_series
                ]
            },
            "options": {
                "responsive": True,
                "scales": {
                    "y": {"beginAtZero": True}
                }
            },
            "export_url": f"/charts/{chart_type}_{hash(str(data_series))}.png"
        }

File: src/groupchat-dotnet/Services/FinancialService.cs
using GroupChat.Dotnet.Models.Financial;

namespace GroupChat.Dotnet.Services;

/// <summary>
/// Core business logic for financial analysis, reporting, and business metrics
/// </summary>
public class FinancialService
{
    /// <summary>
    /// Get sales reports with optional filtering by date range and category
    /// </summary>
    public async Task<List<SalesReport>> GetSalesReportsAsync(string? dateRange = null, string? category = null)
    {
        await Task.Delay(50); // Simulate async operation
        
        var salesData = new List<SalesReport>
        {
            new SalesReport
            {
                ProductId = "PROD001",
                ProductName = "Enterprise Software License",
                Category = "Software",
                SalesAmount = 125000.00m,
                UnitsSold = 50,
                Revenue = 125000.00m,
                ProfitMargin = 0.75,
                SalesDate = DateTime.Now.AddDays(-15),
                Region = "North America",
                SalesRep = "Alice Johnson"
            },
            new SalesReport
            {
                ProductId = "PROD002",
                ProductName = "Consulting Services",
                Category = "Services",
                SalesAmount = 89500.00m,
                UnitsSold = 179,
                Revenue = 89500.00m,
                ProfitMargin = 0.65,
                SalesDate = DateTime.Now.AddDays(-8),
                Region = "Europe",
                SalesRep = "Bob Smith"
            },
            new SalesReport
            {
                ProductId = "PROD003",
                ProductName = "Hardware Components",
                Category = "Hardware",
                SalesAmount = 67800.00m,
                UnitsSold = 226,
                Revenue = 67800.00m,
                ProfitMargin = 0.35,
                SalesDate = DateTime.Now.AddDays(-22),
                Region = "Asia Pacific",
                SalesRep = "Carol Chen"
            },
            new SalesReport
            {
                ProductId = "PROD004",
                ProductName = "Training Package",
                Category = "Services",
                SalesAmount = 34500.00m,
                UnitsSold = 115,
                Revenue = 34500.00m,
                ProfitMargin = 0.80,
                SalesDate = DateTime.Now.AddDays(-5),
                Region = "North America",
                SalesRep = "David Wilson"
            }
        };

        // Apply filters
        if (!string.IsNullOrEmpty(category))
        {
            salesData = salesData.Where(s => s.Category.Contains(category, StringComparison.OrdinalIgnoreCase)).ToList();
        }

        if (!string.IsNullOrEmpty(dateRange))
        {
            // Simple date range filtering - in real implementation this would be more sophisticated
            var cutoffDate = dateRange.ToLower() switch
            {
                "last_week" => DateTime.Now.AddDays(-7),
                "last_month" => DateTime.Now.AddDays(-30),
                "last_quarter" => DateTime.Now.AddDays(-90),
                _ => DateTime.MinValue
            };
            
            if (cutoffDate != DateTime.MinValue)
            {
                salesData = salesData.Where(s => s.SalesDate >= cutoffDate).ToList();
            }
        }

        return salesData;
    }

    /// <summary>
    /// Get revenue data for trend analysis
    /// </summary>
    public async Task<List<RevenueData>> GetRevenueDataAsync(string period)
    {
        await Task.Delay(50);
        
        return new List<RevenueData>
        {
            new RevenueData
            {
                Period = "Q3 2024",
                Revenue = 450000m,
                RecurringRevenue = 320000m,
                OneTimeRevenue = 130000m,
                GrowthRate = 0.15,
                RevenueSource = "Software Sales",
                Department = "Sales"
            },
            new RevenueData
            {
                Period = "Q2 2024",
                Revenue = 390000m,
                RecurringRevenue = 290000m,
                OneTimeRevenue = 100000m,
                GrowthRate = 0.08,
                RevenueSource = "Software Sales",
                Department = "Sales"
            },
            new RevenueData
            {
                Period = "Q1 2024",
                Revenue = 360000m,
                RecurringRevenue = 270000m,
                OneTimeRevenue = 90000m,
                GrowthRate = 0.12,
                RevenueSource = "Software Sales",
                Department = "Sales"
            }
        };
    }

    /// <summary>
    /// Calculate key financial metrics
    /// </summary>
    public async Task<List<FinancialMetrics>> CalculateFinancialMetricsAsync(List<string> metricNames, string period)
    {
        await Task.Delay(50);
        
        var metrics = new List<FinancialMetrics>();
        
        foreach (var metricName in metricNames)
        {
            var metric = metricName.ToLower() switch
            {
                "customer_acquisition_cost" => new FinancialMetrics
                {
                    MetricName = "Customer Acquisition Cost",
                    Value = 85.50m,
                    Period = period,
                    CalculationMethod = "Total Marketing Spend / New Customers Acquired",
                    Benchmark = 100.00m,
                    Trend = TrendDirection.Down,
                    Metadata = new Dictionary<string, object> { ["currency"] = "USD", ["target"] = 75.00 }
                },
                "revenue_growth_rate" => new FinancialMetrics
                {
                    MetricName = "Revenue Growth Rate",
                    Value = 15.3m,
                    Period = period,
                    CalculationMethod = "((Current Period Revenue - Previous Period Revenue) / Previous Period Revenue) * 100",
                    Benchmark = 12.0m,
                    Trend = TrendDirection.Up,
                    Metadata = new Dictionary<string, object> { ["unit"] = "percentage", ["target"] = 18.0 }
                },
                "profit_margin" => new FinancialMetrics
                {
                    MetricName = "Profit Margin",
                    Value = 22.8m,
                    Period = period,
                    CalculationMethod = "(Net Income / Revenue) * 100",
                    Benchmark = 20.0m,
                    Trend = TrendDirection.Up,
                    Metadata = new Dictionary<string, object> { ["unit"] = "percentage", ["industry_avg"] = 18.5 }
                },
                "customer_lifetime_value" => new FinancialMetrics
                {
                    MetricName = "Customer Lifetime Value",
                    Value = 2850.00m,
                    Period = period,
                    CalculationMethod = "Average Revenue Per Customer * Average Customer Lifespan",
                    Benchmark = 2500.00m,
                    Trend = TrendDirection.Up,
                    Metadata = new Dictionary<string, object> { ["currency"] = "USD", ["cohort"] = "enterprise" }
                },
                _ => new FinancialMetrics
                {
                    MetricName = metricName,
                    Value = 0m,
                    Period = period,
                    CalculationMethod = "Unknown metric",
                    Trend = TrendDirection.Stable
                }
            };
            
            metrics.Add(metric);
        }
        
        return metrics;
    }

    /// <summary>
    /// Get budget analysis data
    /// </summary>
    public async Task<List<BudgetAnalysis>> GetBudgetAnalysisAsync(string? department = null)
    {
        await Task.Delay(50);
        
        var budgetData = new List<BudgetAnalysis>
        {
            new BudgetAnalysis
            {
                Department = "Marketing",
                Category = "Digital Advertising",
                BudgetAllocated = 50000m,
                ActualSpending = 47500m,
                Variance = -2500m,
                VariancePercentage = -5.0,
                Period = "Q3 2024",
                Status = "under"
            },
            new BudgetAnalysis
            {
                Department = "Engineering",
                Category = "Software Tools",
                BudgetAllocated = 25000m,
                ActualSpending = 28000m,
                Variance = 3000m,
                VariancePercentage = 12.0,
                Period = "Q3 2024",
                Status = "over"
            },
            new BudgetAnalysis
            {
                Department = "Sales",
                Category = "Travel & Entertainment",
                BudgetAllocated = 15000m,
                ActualSpending = 14800m,
                Variance = -200m,
                VariancePercentage = -1.3,
                Period = "Q3 2024",
                Status = "on-track"
            }
        };

        if (!string.IsNullOrEmpty(department))
        {
            budgetData = budgetData.Where(b => b.Department.Contains(department, StringComparison.OrdinalIgnoreCase)).ToList();
        }

        return budgetData;
    }

    /// <summary>
    /// Get top performing products
    /// </summary>
    public async Task<List<SalesReport>> GetTopPerformingProductsAsync(string timePeriod, string metric = "revenue", int limit = 10)
    {
        var salesData = await GetSalesReportsAsync(timePeriod);
        
        return metric.ToLower() switch
        {
            "revenue" => salesData.OrderByDescending(s => s.Revenue).Take(limit).ToList(),
            "profit_margin" => salesData.OrderByDescending(s => s.ProfitMargin).Take(limit).ToList(),
            "units_sold" => salesData.OrderByDescending(s => s.UnitsSold).Take(limit).ToList(),
            _ => salesData.OrderByDescending(s => s.Revenue).Take(limit).ToList()
        };
    }
}